packages/contracts-bedrock/scripts/DeploySuperchain.s.sol

```diff
@@ -0,0 +1,383 @@
+        require(false, "DeploySuperchain: run is not implemented");
+    }
+
+    // This entrypoint is for use with Solidity tests, where the input and outputs are structs.
+    function run(DeploySuperchainInput.Input memory _input) public returns (DeploySuperchainOutput.Output memory) {
+        // Solidity without file IO, so etch the IO helper contracts.
+        (DeploySuperchainInput dsi, DeploySuperchainOutput dso) = etchIOContracts();
+
+        // Load the input struct into the input contract.
+        dsi.loadInput(_input);
+
+        // Run the deployment script and write outputs to the DeploySuperchainOutput contract.
+        run(dsi, dso);
+
+        // Return the output struct from the output contract.
+        return dso.output();
+    }
+
+    // This entrypoint is useful for testing purposes, as it doesn't use any file I/O.
+    function run(DeploySuperchainInput _dsi, DeploySuperchainOutput _dso) public {
+        // Verify that the input contract has been set.
+        require(_dsi.inputSet(), "DeploySuperchain: Input not set");
+
+        // Deploy the proxy admin, with the owner set to the deployer.
+        deploySuperchainProxyAdmin(_dsi, _dso);
+
+        // Deploy and initialize the superchain contracts.
+        deploySuperchainImplementationContracts(_dsi, _dso);
+        deployAndInitializeSuperchainConfig(_dsi, _dso);
+        deployAndInitializeProtocolVersions(_dsi, _dso);
+
+        // Transfer ownership of the ProxyAdmin from the deployer to the specified owner.
+        transferProxyAdminOwnership(_dsi, _dso);
+
+        // Output assertions, to make sure outputs were assigned correctly.
+        _dso.checkOutput();
+    }
+
+    // -------- Deployment Steps --------
+
+    function deploySuperchainProxyAdmin(DeploySuperchainInput, DeploySuperchainOutput _dso) public {
+        // Deploy the proxy admin, with the owner set to the deployer.
+        // We explicitly specify the deployer as `msg.sender` because for testing we deploy this script from a test
+        // contract. If we provide no argument, the foundry default sender is be the broadcaster during test, but the
+        // broadcaster needs to be the deployer since they are set to the initial proxy admin owner.
+        vm.broadcast(msg.sender);
+        ProxyAdmin superchainProxyAdmin = new ProxyAdmin(msg.sender);
+
+        vm.label(address(superchainProxyAdmin), "SuperchainProxyAdmin");
+        _dso.set(_dso.superchainProxyAdmin.selector, address(superchainProxyAdmin));
+    }
+
+    function deploySuperchainImplementationContracts(DeploySuperchainInput, DeploySuperchainOutput _dso) public {
+        // Deploy implementation contracts.
+        vm.startBroadcast(msg.sender);
+        SuperchainConfig superchainConfigImpl = new SuperchainConfig();
+        ProtocolVersions protocolVersionsImpl = new ProtocolVersions();
+        vm.stopBroadcast();
+
+        vm.label(address(superchainConfigImpl), "SuperchainConfigImpl");
```

Comment:
We dont use the `Impl` suffix in the other scripts, we did use it a couple of refactors ago

Reply:
I added it it here because the explicitness makes it clear what the purpose is, especially in #11539 where some contract deployments are the implementation contracts, and some are singletons (i.e. they are not proxied). So the `Impl` suffix gives us a way to easily convey the purpose of each contract to avoid footguns due to ambiguous names